File: src/main/kotlin/xyz/deseteral/lebkuchenfm/domain/commands/parser/TextCommandParser.kt
package xyz.deseteral.lebkuchenfm.domain.commands.parser

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import xyz.deseteral.lebkuchenfm.domain.commands.model.Command

@Component
class TextCommandParser(@Value("\${COMMAND_PROMPT:/fm}") val commandPrompt: String) {
    fun parse(text: String): Command {
        val tokens = text.split(" ")
            .map { it.trim() }
            .filter { it.isNotEmpty() }

        if (tokens.size < 2) throw TextIsNotACommandException(text)

        val prompt = tokens.first()
        val key = tokens[1]

        if (prompt != commandPrompt) throw TextIsNotACommandException(text)

        val rawArgsIndex = (text.indexOf(key) + (key.length) + 1)
        val rawArgs = try {
            text.substring(rawArgsIndex)
        } catch (ex: Exception) {
            ""
        }

        return Command(key, rawArgs)
    }
}

File: packages/service/src/domain/commands/text-command-parser.ts
import { Command } from '@service/domain/commands/model/command';
import { Configuration } from '@service/infrastructure/configuration';
import { Service } from 'typedi';

@Service()
class TextCommandParser {
  constructor(private configuration: Configuration) { }

  public parseTextToCommand(text: string): (Command | null) {
    const tokens = text.split(' ')
      .map((s) => s.trim())
      .filter((s) => (s.length > 0));

    if (tokens.length < 2) return null;

    const prompt = tokens[0];
    const key = tokens[1];

    if (prompt !== this.configuration.COMMAND_PROMPT) return null;

    const rawArgsIndex = (text.indexOf(key) + key.length + 1);
    const rawArgs = text.substring(rawArgsIndex);

    return new Command(key, rawArgs);
  }
}

export { TextCommandParser };
